File: backend/app/api/routes/projects.py

from decimal import Decimal

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import case, func
from sqlalchemy.orm import Session

from app.api.deps import PaginationParams
from app.db.database import get_db
from app.models.client import Client
from app.models.enums import ProjectStatus, TimeEntryStatus
from app.models.project import Project
from app.models.time_entry import TimeEntry
from app.schemas.common import Page
from app.schemas.project import (
    ProjectCreate,
    ProjectDetailRead,
    ProjectRead,
    ProjectStats,
    ProjectUpdate,
)

router = APIRouter()


def _get_or_404(project_id: int, db: Session) -> Project:
    project = db.get(Project, project_id)
    if project is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Проект с id={project_id} не найден",
        )
    return project


def _compute_stats(project_id: int, db: Session) -> ProjectStats:
    zero = Decimal("0")
    row = db.query(
        func.coalesce(func.sum(TimeEntry.duration_hours), zero).label("total_hours"),
        func.coalesce(
            func.sum(
                case(
                    (TimeEntry.status == TimeEntryStatus.confirmed, TimeEntry.duration_hours),
                    else_=zero,
                )
            ),
            zero,
        ).label("confirmed_hours"),
        func.coalesce(
            func.sum(
                case(
                    (TimeEntry.status != TimeEntryStatus.billed, TimeEntry.duration_hours),
                    else_=zero,
                )
            ),
            zero,
        ).label("unbilled_hours"),
    ).filter(TimeEntry.project_id == project_id).one()

    return ProjectStats(
        total_hours=row.total_hours,
        confirmed_hours=row.confirmed_hours,
        unbilled_hours=row.unbilled_hours,
    )


@router.get(
    "",
    response_model=Page[ProjectRead],
    summary="Список проектов",
)
def list_projects(
    client_id: int | None = Query(None, description="Фильтр по клиенту"),
    status_filter: ProjectStatus | None = Query(None, alias="status", description="Фильтр по статусу"),
    pagination: PaginationParams = Depends(),
    db: Session = Depends(get_db),
) -> Page[ProjectRead]:
    q = db.query(Project)
    if client_id is not None:
        q = q.filter(Project.client_id == client_id)
    if status_filter is not None:
        q = q.filter(Project.status == status_filter)
    total = q.count()
    items = (
        q.order_by(Project.created_at.desc())
        .offset(pagination.offset)
        .limit(pagination.size)
        .all()
    )
    return Page.create(items=items, total=total, page=pagination.page, size=pagination.size)


@router.post(
    "",
    response_model=ProjectRead,
    status_code=status.HTTP_201_CREATED,
    summary="Создать проект",
    responses={404: {"description": "Клиент не найден"}},
)
def create_project(data: ProjectCreate, db: Session = Depends(get_db)) -> Project:
    if db.get(Client, data.client_id) is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Клиент с id={data.client_id} не найден",
        )
    project = Project(**data.model_dump())
    db.add(project)
    db.commit()
    db.refresh(project)
    return project


@router.get(
    "/{project_id}",
    response_model=ProjectDetailRead,
    summary="Получить проект с статистикой",
    responses={404: {"description": "Проект не найден"}},
)
def get_project(project_id: int, db: Session = Depends(get_db)) -> ProjectDetailRead:
    project = _get_or_404(project_id, db)
    stats = _compute_stats(project_id, db)
    return ProjectDetailRead.model_validate(
        {**project.__dict__, "stats": stats}
    )


@router.put(
    "/{project_id}",
    response_model=ProjectRead,
    summary="Обновить проект",
    responses={404: {"description": "Проект не найден"}},
)
def update_project(
    project_id: int,
    data: ProjectUpdate,
    db: Session = Depends(get_db),
) -> Project:
    project = _get_or_404(project_id, db)
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(project, field, value)
    db.commit()
    db.refresh(project)
    return project


@router.delete(
    "/{project_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    summary="Удалить проект",
    description="Проект удаляется только при отсутствии записей времени.",
    responses={
        404: {"description": "Проект не найден"},
        409: {"description": "У проекта есть записи времени"},
    },
)
def delete_project(project_id: int, db: Session = Depends(get_db)) -> None:
    project = _get_or_404(project_id, db)

    entry_count = (
        db.query(func.count(TimeEntry.id))
        .filter(TimeEntry.project_id == project_id)
        .scalar()
    )
    if entry_count:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=(
                f"Невозможно удалить проект: существует {entry_count} запис(ь/и/ей) времени. "
                "Сначала удалите записи."
            ),
        )

    db.delete(project)
    db.commit()

File: frontend/src/services/projects.ts
import { supabase } from '@/lib/supabase'
import type { Page, Project, ProjectCreate, ProjectDetail, ProjectUpdate, ProjectStatus } from '@/types'

const TABLE = 'projects'

function paginate<T>(items: T[], total: number, page: number, size: number): Page<T> {
  return { items, total, page, size, pages: Math.ceil(total / size) || 1 }
}

export const projectsService = {
  list: async (params: { client_id?: number; status?: ProjectStatus; page?: number; size?: number }): Promise<Page<Project>> => {
    const page = params.page ?? 1
    const size = params.size ?? 20
    const from = (page - 1) * size
    const to = from + size - 1

    let query = supabase.from(TABLE).select('*', { count: 'exact' })
    if (params.client_id) query = query.eq('client_id', params.client_id)
    if (params.status) query = query.eq('status', params.status)
    query = query.order('created_at', { ascending: false }).range(from, to)

    const { data, error, count } = await query
    if (error) throw error
    return paginate(data as Project[], count ?? 0, page, size)
  },

  get: async (id: number): Promise<ProjectDetail> => {
    const { data: project, error } = await supabase.from(TABLE).select('*').eq('id', id).single()
    if (error) throw new Error('Проект не найден')

    const { data: entries } = await supabase
      .from('time_entries')
      .select('duration_hours, status')
      .eq('project_id', id)

    const all = entries ?? []
    const total_hours = all.reduce((s, e) => s + parseFloat(e.duration_hours), 0)
    const confirmed_hours = all.filter(e => e.status !== 'draft').reduce((s, e) => s + parseFloat(e.duration_hours), 0)
    const unbilled_hours = all.filter(e => e.status === 'confirmed').reduce((s, e) => s + parseFloat(e.duration_hours), 0)

    return {
      ...(project as Project),
      stats: {
        total_hours: total_hours.toFixed(2),
        confirmed_hours: confirmed_hours.toFixed(2),
        unbilled_hours: unbilled_hours.toFixed(2),
      },
    }
  },

  create: async (data: ProjectCreate): Promise<Project> => {
    const { data: created, error } = await supabase.from(TABLE).insert({
      client_id: data.client_id,
      name: data.name,
      description: data.description ?? null,
      hourly_rate: data.hourly_rate ?? null,
      currency: data.currency ?? 'RUB',
      status: data.status ?? 'active',
    }).select().single()
    if (error) throw error
    return created as Project
  },

  update: async (id: number, data: ProjectUpdate): Promise<Project> => {
    const { data: updated, error } = await supabase.from(TABLE).update(data).eq('id', id).select().single()
    if (error) throw new Error('Проект не найден')
    return updated as Project
  },

  delete: async (id: number): Promise<void> => {
    const { error } = await supabase.from(TABLE).delete().eq('id', id)
    if (error) throw error
  },
}
